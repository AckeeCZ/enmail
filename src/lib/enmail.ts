export interface Message {
    from?: string;
    to?: string | string[];
    content?: string;
    subject?: string;
}

export type SendMessage<AdapterOptions = any, SendOptions = any> =
    (message: Message, adapter: AdapterSend | Adapter<AdapterOptions, SendOptions>, sendOptions?: SendOptions) => Promise<unknown>;
export type AdapterSend<SendOptions = any> = (message: Message, options?: SendOptions) => Promise<unknown>;

export interface Adapter<AdapterOptions = any, SendOptions = any> {
    options: AdapterOptions;
    mailService: any;
    sender: (sendOptions?: SendOptions) => AdapterSend<SendOptions>;
}
