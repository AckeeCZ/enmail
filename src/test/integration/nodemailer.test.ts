import enmail from '../../index';

jest.mock('nodemailer');
const nodemailer = require('nodemailer');
nodemailer.createTransport.mockReturnValue({ sendMail: jest.fn(data => data) });

describe('Nodemailer adapter', () => {
    const getGmailAdapter = () => enmail.createAdapter.nodemailer(options);
    const options = {
        service: 'gmail',
        auth: {
            type: 'login' as const,
            user: 'test2@example.com',
            pass: '********',
        },
    };
    const adapter = getGmailAdapter();
    const sendOptions = { bcc: 'test3@example.com' };
    const body = {
        to: 'test@example.com',
        from: 'test2@example.com',
        content: 'Hello world',
        subject: 'Hello',
    };
    it('Gmail adapter can be created', () => {
        expect(typeof adapter.sender).toEqual('function');
        expect(adapter.options).toStrictEqual(options);
        expect(typeof adapter.mailService.sendMail).toEqual('function');
        expect(adapter).toMatchSnapshot();
    });
    it('Sender will be successfully returned', () => {
        const sender = enmail.sender(adapter);
        expect(typeof sender).toEqual('function');
        expect(sender).toEqual(adapter.sender);
    });
    it('An email is created via adapter', async () => {
        const gmailMessage = await adapter.sender()(body, sendOptions) as any;
        expectNodemailerEmailObj(gmailMessage, ['bcc']);
    });
    it('An email is created via adapter with send options', async () => {
        const gmailMessage = await adapter.sender(sendOptions)(body) as any;
        expectNodemailerEmailObj(gmailMessage, ['bcc']);
    });
    it('An email is created via adapter without send options', async () => {
        const gmailMessage = await adapter.sender()(body) as any;
        expectNodemailerEmailObj(gmailMessage);
    });
    it('An email is created via enmail.send', async () => {
        const gmailMessage = await enmail.send(body, adapter) as any;
        expectNodemailerEmailObj(gmailMessage);
    });
    it('An email is created via enmail.send with sendOptions', async () => {
        const gmailMessage = await enmail.send(body, adapter, sendOptions) as any;
        expectNodemailerEmailObj(gmailMessage, ['bcc']);
    });
    it('An email is created via sender', async () => {
        const gmailMessage = await enmail.send(body, adapter.sender()) as any;
        expectNodemailerEmailObj(gmailMessage);
    });
    it('An email is created via mailService', async () => {
        const { content, ...rest } = body;
        const gmailMessage = await adapter.mailService.sendMail({ ...rest, html: content }) as any;
        expectNodemailerEmailObj(gmailMessage);
    });
});

const expectNodemailerEmailObj = (email: any, additionalEmailProps?: string[]) => {
    const props = ['to', 'html', 'from', 'subject', ...(additionalEmailProps || [])]
        .filter(x => x)
        .sort();
    const messageProps = Object.keys(email).sort();
    expect(messageProps).toEqual(props);
};
