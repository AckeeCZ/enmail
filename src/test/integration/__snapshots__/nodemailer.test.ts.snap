// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nodemailer adapter Gmail adapter can be created 1`] = `
Object {
  "mailService": Object {
    "sendMail": [MockFunction],
  },
  "options": Object {
    "auth": Object {
      "pass": "********",
      "type": "login",
      "user": "test2@example.com",
    },
    "service": "gmail",
  },
  "sender": [Function],
}
`;
